src/window/gl_canvas.rs

```diff
@@ -82,86 +87,83 @@ impl AbstractCanvas for GLCanvas {
-                } => {
-                    let (x, y) = match delta {
-                        glutin::MouseScrollDelta::LineDelta(dx, dy) => {
-                            (dx as f64 * 10.0, dy as f64 * 10.0)
-                        }
-                        glutin::MouseScrollDelta::PixelDelta(delta) => delta.into(),
-                    };
-                    let modifiers = translate_modifiers(modifiers);
-                    let _ = out_events.send(WindowEvent::Scroll(x, y, modifiers));
-                }
-                glutin::WindowEvent::KeyboardInput { input, .. } => {
-                    let action = translate_action(input.state);
-                    let key = translate_key(input.virtual_keycode);
-                    let modifiers = translate_modifiers(input.modifiers);
-                    key_states[key as usize] = action;
-                    let _ = out_events.send(WindowEvent::Key(key, action, modifiers));
-                }
-                glutin::WindowEvent::ReceivedCharacter(c) => {
-                    let _ = out_events.send(WindowEvent::Char(c));
-                }
+                        let _ = out_events.send(WindowEvent::Touch(
+                            touch.id,
+                            touch.location.x,
+                            touch.location.y,
+                            action,
+                            Modifiers::empty(),
+                        ));
+                    }
+                    glutin::event::WindowEvent::MouseWheel {
+                        delta, modifiers, ..
+                    } => {
+                        let (x, y) = match delta {
+                            glutin::event::MouseScrollDelta::LineDelta(dx, dy) => {
+                                (dx as f64 * 10.0, dy as f64 * 10.0)
+                            }
+                            glutin::event::MouseScrollDelta::PixelDelta(delta) => delta.into(),
+                        };
+                        let modifiers = translate_modifiers(modifiers);
+                        let _ = out_events.send(WindowEvent::Scroll(x, y, modifiers));
+                    }
+                    glutin::event::WindowEvent::KeyboardInput { input, .. } => {
+                        let action = translate_action(input.state);
+                        let key = translate_key(input.virtual_keycode);
+                        let modifiers = translate_modifiers(input.modifiers);
+                        key_states[key as usize] = action;
+                        let _ = out_events.send(WindowEvent::Key(key, action, modifiers));
+                    }
+                    glutin::event::WindowEvent::ReceivedCharacter(c) => {
+                        let _ = out_events.send(WindowEvent::Char(c));
+                    }
+                    _ => {}
+                },
                 _ => {}
-            },
-            _ => {}
+            };
+            *control_flow = ControlFlow::Exit;
```

Comment:
I don't believe the use of `run_return` here is correct... is it? This could very well end up discarding events during window resizing on Windows and macOS. Also, given that `winit` starts an event iteration with `Event::NewEvents`, by exiting the loop on the first event received, how does this actually receive any events at all? Not to mention the possibility of multiple events being emitted consecutively.

Reply:
I tested it on Arch Linux and the examples run successfully. Your concern is true. Frankly I do not know much about the events ([doc](https://docs.rs/glutin/0.25.1/glutin/event/enum.Event.html)). I'll learn more about that and you can give some instructions.

Reply:
Seems `Event::RedrawEventsCleared` marks the end of all input and rendering events in the doc. Correct me if I'm wrong.